File: src/com/easy/DiameterofBinaryTree_543.java
/**
 * 
 */
package com.easy;

/**
 * @author Chris Lin
 * @version 1.1
 */
public class DiameterofBinaryTree_543 {
    public int mResult = 0;

    // Use Recursive
    // Time Complexity: O(n)
    // Space Complexity:O(1)
    // Runtime: 0 ms, faster than 100.00%
    // Memory Usage: 37.7 MB, less than 94.26%
    public int diameterOfBinaryTree(TreeNode root) {
        maxDepth(root);

        return mResult;
    }

    private int maxDepth(TreeNode root) {
        if (root == null) {
            return 0;
        }
        int mDepthLeft = maxDepth(root.left);
        int mDepthRight = maxDepth(root.right);
        mResult = Math.max(mDepthLeft + mDepthRight, mResult);

        return Math.max(mDepthLeft, mDepthRight) + 1;
    }

    // Use Recursive without global variable
    // Time Complexity: O(n)
    // Space Complexity:O(1)
    // Runtime: 10 ms, faster than 7.69%
    // Memory Usage: 38.5 MB, less than 36.36% 
    public int diameterOfBinaryTree2(TreeNode root) {
        if (root == null) {
            return 0;
        }
        int mDepthLeft = getDepth(root.left);
        int mDepthRight = getDepth(root.right);

        int mDiameterLeft = diameterOfBinaryTree2(root.left);
        int mDiameterRight = diameterOfBinaryTree2(root.right);

        return Math.max(mDepthLeft + mDepthRight, Math.max(mDiameterLeft, mDiameterRight));
    }

    private int getDepth(TreeNode root) {
        return (root == null) ? 0 : Math.max(getDepth(root.left), getDepth(root.right)) + 1;
    }
}

File: Golang_src/easy/diameterOfBinaryTree_543.go
package easy

// TreeNode for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// Use DFS:
// Time Complexity: O(n)
// Space Complexity:O(n)
// Runtime: 4 ms, faster than 94.69%
// Memory Usage: 4.5 MB, less than 100.00%
func diameterOfBinaryTree(root *TreeNode) int {
	result := 0

	depth(root, &result)

	return result
}

func depth(root *TreeNode, result *int) int {
	if root == nil {
		return 0
	}
	l := depth(root.Left, result)
	r := depth(root.Right, result)

	*result = max((l + r), *result)

	return 1 + max(l, r)
}

func max(a, b int) int {
	if a > b {
		return a
	}

	return b
}
